CRM/Contribute/Form/Contribution/Main.php

```diff
@@ -1310,20 +1310,30 @@ public function assignFormVariablesByContributionID() {
       return;
     }
     if (!$this->getContactID()) {
-      CRM_Core_Error::statusBounce(ts("Returning since there is no contact attached to this contribution id."));
+      throw new CRM_Core_Exception(ts("Returning since there is no contact attached to this contribution id."));
```

Comment:
I'm a bit torn here - I think it is less rude to bounce them?

Reply:
I think previous statusBounce was correct as per the Error message which says *Returning*.. Or I would suggest:
```
CRM_Core_Error::statusBounce(ts("There is no contact attached to this contribution id."));
```

Reply:
Hmm, statusBounce is what was used before. This was modified to satisfy @magnolia61 requirement [in this comment](https://github.com/civicrm/civicrm-core/pull/12319#issuecomment-398029244) of not seeing any error when the user is navigated to CMS home page after the bounce is triggered. I'm

Reply:
It was because a message to users clicking a Pay Now checksum link which was already paid or expired got no message at all. With all improvement ideas, I think it could be a separate issue to provide a end-user friendly error page for non-logged users.